msdk-io/msdk-io-mzml2/src/main/java/io/github/msdk/io/mzml2/MzMLSpectrum.java

```diff
@@ -154,95 +148,153 @@ public Integer getNumberOfDataPoints() {
 
       compressions.add(getIntensityBinaryDataInfo().getCompressionType());
 
-      InputStream encodedIs = new ByteBufferInputStreamAdapter(mappedByteBufferInputStream, 0,
+      InputStream encodedIs = new ByteBufferInputStreamAdapter(mappedByteBufferInputStream,
+          getIntensityBinaryDataInfo().getPosition(),
           getIntensityBinaryDataInfo().getEncodedLength());
       InputStream decodedIs = Base64.getDecoder().wrap(encodedIs);
       byte[] decodedData = IOUtils.toByteArray(decodedIs);
+      System.out.println(new String(IOUtils.toByteArray(encodedIs)));
 
-      result = MzMLIntensityPeaksDecoder.decode(decodedData, 0, precision,
+      intensityValues = MzMLIntensityPeaksDecoder.decode(decodedData, decodedData.length, precision,
           getIntensityBinaryDataInfo().getArrayLength(), compressions).arr;
     } catch (Exception e) {
       throw (new MSDKRuntimeException(e));
     }
 
-    return result;
+    return intensityValues;
   }
 
   @Override
-  public Float getTIC() {
-    // TODO Auto-generated method stub
-    return null;
+  public MsSpectrumType getSpectrumType() {
+    MsSpectrumType spectrumType = SpectrumTypeDetectionAlgorithm.detectSpectrumType(mzValues,
+        intensityValues, getMzBinaryDataInfo().getArrayLength());
+    return spectrumType;
   }
 
   @Override
-  public Range<Double> getMzRange() {
-    // TODO Auto-generated method stub
-    return null;
+  public Float getTIC() {
+    Float tic = MsSpectrumUtil.getTIC(intensityValues, getMzBinaryDataInfo().getArrayLength());
+    return tic;
   }
 
   @Override
-  public MzTolerance getMzTolerance() {
-    // TODO Auto-generated method stub
-    return null;
+  public Range<Double> getMzRange() {
+    Range<Double> mzRange =
+        MsSpectrumUtil.getMzRange(mzValues, getMzBinaryDataInfo().getArrayLength());
```

Comment:
There are CV terms for mz range:
````
<cvParam cvRef="MS" accession="MS:1000528" name="lowest observed m/z" value="116.825372650742" unitCvRef="MS" unitAccession="MS:1000040" unitName="m/z"/>
          <cvParam cvRef="MS" accession="MS:1000527" name="highest observed m/z" value="1767.680685613041" unitCvRef="MS" unitAccession="MS:1000040" unitName="m/z"/>
````
Use the MsSpectrumUtil method if these are not present in the mzML file. Save the result for reuse (same as with TIC).